### DISCUSSION TITLE
Can you split an input to an interconnected system?

### DISCUSSION THREAD

**Discussion by corbinklett** [python-control/python-control#1047] [Q&A] [answered]
If I have an input that is external to an interconnected system, and I want this external input to connect into multiple systems contained within, the only way I can think to do this is by using `set_input_map`. Is there a more elegant way? Even when I use `set_input_map`, I still get the "unused input" warning. 

Part of the motivation for this question is that it is difficult to understand how an external input is being used if you can only specify it in terms of the subsystem that it connects to. It might be intended for the input to connect to multiple subsystems.

Here is an example of how I have implemented:

```python
import numpy as np
import control as ct

def dynamics(t, x, u, p):
    A = np.array([[0, 1], [-p['spring_constant'], -p['damping_coefficient']]])
    B = np.array([[0], [1]])
    return A @ x + B @ u

def outputs(t, x, u, p):
    C = np.array([[1, 0]])
    D = np.array([[0]])
    return C @ x + D @ u

sys1 = ct.nlsys(updfcn=dynamics, outfcn=outputs, states=['pos', 'vel'], inputs=['u1'], outputs=['pos'], name='sys1')
sys1.params = {'spring_constant': 10, 'damping_coefficient': 0.5}

sys2 = ct.nlsys(updfcn=dynamics, outfcn=outputs, states=['pos', 'vel'], inputs=['u2'], outputs=['pos'], name='sys2')
sys2.params = {'spring_constant': 1, 'damping_coefficient': 1}

sys = ct.interconnect(
    (sys1, sys2),
    # inplist=['u'], # only works if both subsystem inputs are named u
    outlist=['sys1.pos', 'sys2.pos'],
    outputs=['pos1', 'pos2']
)

# /venv/lib/python3.12/site-packages/control/nlsys.py:1192: UserWarning: Unused input(s) in InterconnectedSystem: (1, 0)=sys2.u2; (0, 0)=sys1.u1

sys.set_input_map(np.array([[1],[1]]))
sys.set_inputs('external_input')

t = np.linspace(0, 10, 1000)

u0 = lambda t: 1
U = np.array([
    [u0(ti) for ti in t]
])

results = ct.input_output_response(sys, T=t, U=U)

p = results.plot()
p.figure.show()

```

**Comment by corbinklett** [comment]
Would be nice to be able to do something like this:

```python
sys = ct.interconnect(
    (sys1, sys2),
    connections=(
        ['sys1.u1', 'external_input'],
        ['sys2.u2', 'external_input'],
    ),
    inplist=['external_input'],
    outlist=['sys1.pos', 'sys2.pos'],
    outputs=['pos1', 'pos2'],
)
```

**Comment by bavcol** [comment]
Hi corbin,
from your example:
`    # inplist=['u'], # only works if both subsystem inputs are named u`
The signal used as input of sys1 and sys2 is _u_, so that is the correct input name for sys1 and sys2. No need to introduce _u1_ and _u2_, when _u_ == _u1_ == _u2_.
Also, it would make sense to me to use different names for the outputs of sys1 and sys2, e.g. _pos1_ and _pos2_ instead of the same name _pos_ for different signals.

Or put in different words: I recomment using using the same name for the same signal and different names for different signals to keep things simple.

Note: I am not an python-control expert, so please let me know if I am missing something.

**Reply by corbinklett** [reply]
Your statements are all correct. In my case, though, I may be re-using simulation components where I don't have control over the naming, but I do know that I want to split the external input and connect it to multiple components. But maybe there's a simple way to modify my approach to align it with the current conventions.

**Comment by murrayrm** [comment, marked as answer]
You can accomplish this interconnection using the following:
```
    sys = ct.interconnect(
        [sys1, sys2],
        connections=None,
        inplist=[
            ['sys1.u1', 'sys2.u2']      # Input connects to multiple signals
        ],
        inputs='external_input',
        outlist=['sys1.pos', 'sys2.pos'],
        outputs=['pos1', 'pos2'],
    )    
```
The reason this works is that if you specify a list as an entry for `inplist`, it will connect that input (`external_input`) here to all of the signals in that list.

I also like the alternative syntax you have proposed above, so I've added issue #1049 to track its implementation.
